File: js/test/redaction.conformance.test.mjs
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import test from 'node:test';
import { fileURLToPath } from 'node:url';
import { createSecretRedactionSanitizer, redactSecretText, redactSecretTextLightweight } from '../.test-dist/index.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const fixturesDir = join(__dirname, '../../redaction/fixtures/');

/** Loads a shared fixture file. Every redaction engine reads the same path. */
export function loadFixture(name) {
  return JSON.parse(readFileSync(join(fixturesDir, name), 'utf8'));
}

const stringCases = loadFixture('strings.json').cases;

test('conformance: redaction strings', async (t) => {
  assert.ok(stringCases.length > 0, 'strings.json has no cases');

  for (const testCase of stringCases) {
    await t.test(testCase.id, () => {
      const options = { redactEmailAddresses: testCase.emails };
      let actual;
      if (testCase.mode === 'full') {
        actual = redactSecretText(testCase.input, options);
      } else if (testCase.mode === 'light') {
        actual = redactSecretTextLightweight(testCase.input, options);
      } else {
        throw new Error(`unknown mode ${testCase.mode}`);
      }
      assert.equal(actual, testCase.expected);

      // Redacting a JSON payload must not corrupt its structure. Tool call
      // arguments reach the sanitizer as raw JSON bytes, and a truncated
      // payload is unreadable downstream.
      if (parsesAsJson(testCase.input)) {
        assert.ok(parsesAsJson(actual), `redacting ${testCase.id} produced invalid JSON: ${actual}`);
      }
    });
  }
});

function parsesAsJson(value) {
  if (!value.startsWith('{') && !value.startsWith('[')) return false;
  try {
    JSON.parse(value);
    return true;
  } catch {
    return false;
  }
}

const generationFixtures = loadFixture('generations.json');
const probe = generationFixtures.probe;

const assistantParts = (value, callId) => [
  { type: 'text', text: value },
  { type: 'thinking', thinking: value },
  { type: 'tool_call', toolCall: { id: callId, name: 'bash', inputJSON: value } },
];

const toolParts = (value, callId) => [
  { type: 'text', text: value },
  { type: 'tool_result', toolResult: { toolCallId: callId, name: 'bash', content: value, contentJSON: value } },
];

/** Fills every slot in the matrix with the same probe. */
function buildProbeGeneration(value) {
  return {
    id: 'gen-conformance',
    mode: 'SYNC',
    operationName: 'generateText',
    model: { provider: 'openai', name: 'gpt-5' },
    startedAt: new Date('2026-01-01T00:00:00Z'),
    completedAt: new Date('2026-01-01T00:00:01Z'),
    systemPrompt: value,
    conversationTitle: value,
    callError: value,
    input: [
      { role: 'user', parts: [{ type: 'text', text: value }] },
      { role: 'assistant', parts: assistantParts(value, 'call-1') },
      { role: 'tool', parts: toolParts(value, 'call-1') },
    ],
    output: [
      { role: 'assistant', parts: assistantParts(value, 'call-2') },
      { role: 'tool', parts: toolParts(value, 'call-2') },
    ],
  };
}

function slotValues(generation) {
  return {
    systemPrompt: generation.systemPrompt,
    conversationTitle: generation.conversationTitle,
    callError: generation.callError,
    'input.user.text': generation.input[0].parts[0].text,
    'input.assistant.text': generation.input[1].parts[0].text,
    'input.assistant.thinking': generation.input[1].parts[1].thinking,
    'input.assistant.toolCallInputJson': generation.input[1].parts[2].toolCall.inputJSON,
    'input.tool.text': generation.input[2].parts[0].text,
    'input.tool.toolResultContent': generation.input[2].parts[1].toolResult.content,
    'input.tool.toolResultContentJson': generation.input[2].parts[1].toolResult.contentJSON,
    'output.assistant.text': generation.output[0].parts[0].text,
    'output.assistant.thinking': generation.output[0].parts[1].thinking,
    'output.assistant.toolCallInputJson': generation.output[0].parts[2].toolCall.inputJSON,
    'output.tool.text': generation.output[1].parts[0].text,
    'output.tool.toolResultContent': generation.output[1].parts[1].toolResult.content,
    'output.tool.toolResultContentJson': generation.output[1].parts[1].toolResult.contentJSON,
  };
}

test('conformance: redaction generation slots', async (t) => {
  for (const testCase of generationFixtures.cases) {
    await t.test(testCase.id, () => {
      const sanitize = createSecretRedactionSanitizer(
        {
          redactInputMessages: testCase.redactInputMessages,
          redactEmailAddresses: testCase.redactEmailAddresses,
        },
        {},
      );
      const actual = slotValues(sanitize(buildProbeGeneration(probe.input)));

      const expected = {};
      for (const [slot, mode] of Object.entries(testCase.slots)) {
        expected[slot] = probe[mode];
      }

      assert.deepEqual(
        Object.keys(actual).sort(),
        Object.keys(expected).sort(),
        'JS harness slots and fixture slots disagree',
      );
      assert.deepEqual(actual, expected);
    });
  }
});

File: go/agento11y/redaction_conformance_test.go
package agento11y

import (
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"testing"
)

// stringFixtures mirrors redaction/fixtures/strings.json, the shared case file
// every redaction engine loads. Keeping the shape here rather than in a helper
// package keeps the SDK free of test-only dependencies.
type stringFixtures struct {
	Cases []stringCase `json:"cases"`
}

type stringCase struct {
	ID       string `json:"id"`
	Mode     string `json:"mode"`
	Emails   bool   `json:"emails"`
	Input    string `json:"input"`
	Expected string `json:"expected"`
}

func loadStringFixtures(t *testing.T) stringFixtures {
	t.Helper()
	path := filepath.Join("..", "..", "redaction", "fixtures", "strings.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var fixtures stringFixtures
	if err := json.Unmarshal(raw, &fixtures); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}
	if len(fixtures.Cases) == 0 {
		t.Fatalf("%s has no cases", path)
	}
	return fixtures
}

func TestConformanceRedactionStrings(t *testing.T) {
	for _, tc := range loadStringFixtures(t).Cases {
		t.Run(tc.ID, func(t *testing.T) {
			var got string
			switch tc.Mode {
			case "full":
				got = redactFull(tc.Input, tc.Emails)
			case "light":
				got = redactLight(tc.Input, tc.Emails)
			default:
				t.Fatalf("unknown mode %q", tc.Mode)
			}
			if got != tc.Expected {
				t.Errorf("mode=%s emails=%v\n input: %q\n   got: %q\n  want: %q", tc.Mode, tc.Emails, tc.Input, got, tc.Expected)
			}
		})
	}
}

// TestConformanceRedactionFixtureCoverage keeps the shared fixture file honest:
// a pattern with no positive case can be mis-escaped in one language and still
// pass every suite.
func TestConformanceRedactionFixtureCoverage(t *testing.T) {
	cases := loadStringFixtures(t).Cases

	seen := map[string]map[string]bool{}
	for _, tc := range cases {
		for _, id := range markerIDs(tc.Expected) {
			if seen[id] == nil {
				seen[id] = map[string]bool{}
			}
			seen[id][tc.Mode] = true
		}
	}

	for _, p := range tier1Patterns {
		if !seen[p.id]["full"] || !seen[p.id]["light"] {
			t.Errorf("tier 1 pattern %q needs a positive fixture in both full and light mode", p.id)
		}
	}
	if !seen[emailPattern.id]["full"] && !seen[emailPattern.id]["light"] {
		t.Errorf("email pattern needs a positive fixture")
	}
	for _, p := range tier2Patterns {
		if !seen[p.id]["full"] {
			t.Errorf("tier 2 pattern %q needs a positive fixture in full mode", p.id)
		}
	}
}

var markerPattern = regexp.MustCompile(`\[REDACTED:([a-z0-9-]+)\]`)

// markerIDs extracts the pattern ids from the `[REDACTED:<id>]` markers in an
// expected value.
func markerIDs(expected string) []string {
	var ids []string
	for _, match := range markerPattern.FindAllStringSubmatch(expected, -1) {
		ids = append(ids, match[1])
	}
	return ids
}

// generationFixtures mirrors redaction/fixtures/generations.json.
type generationFixtures struct {
	Probe map[string]string `json:"probe"`
	Cases []generationCase  `json:"cases"`
}

type generationCase struct {
	ID                   string            `json:"id"`
	RedactInputMessages  bool              `json:"redactInputMessages"`
	RedactEmailAddresses bool              `json:"redactEmailAddresses"`
	Slots                map[string]string `json:"slots"`
}

func loadGenerationFixtures(t *testing.T) generationFixtures {
	t.Helper()
	path := filepath.Join("..", "..", "redaction", "fixtures", "generations.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var fixtures generationFixtures
	if err := json.Unmarshal(raw, &fixtures); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}
	if len(fixtures.Cases) == 0 {
		t.Fatalf("%s has no cases", path)
	}
	return fixtures
}

// buildProbeGeneration fills every slot in the matrix with the same probe so
// the sanitizer's per-slot mode is what the assertions read back. The media
// part is Go-only and has no slot in the shared matrix; the test asserts on it
// directly.
func buildProbeGeneration(probe string) Generation {
	toolParts := func() []Part {
		return []Part{
			{Kind: PartKindText, Text: probe},
			{Kind: PartKindToolResult, ToolResult: &ToolResult{
				Name:        "bash",
				Content:     probe,
				ContentJSON: []byte(probe),
			}},
		}
	}
	assistantParts := func() []Part {
		return []Part{
			{Kind: PartKindText, Text: probe},
			{Kind: PartKindThinking, Thinking: probe},
			{Kind: PartKindToolCall, ToolCall: &ToolCall{Name: "bash", InputJSON: []byte(probe)}},
		}
	}

	return Generation{
		SystemPrompt:      probe,
		ConversationTitle: probe,
		CallError:         probe,
		Input: []Message{
			{Role: RoleUser, Parts: []Part{
				{Kind: PartKindText, Text: probe},
				{Kind: PartKindMedia, Media: &Media{Kind: "image", URL: probe}},
			}},
			{Role: RoleAssistant, Parts: assistantParts()},
			{Role: RoleTool, Parts: toolParts()},
		},
		Output: []Message{
			{Role: RoleAssistant, Parts: assistantParts()},
			{Role: RoleTool, Parts: toolParts()},
		},
	}
}

func generationSlotValues(g Generation) map[string]string {
	return map[string]string{
		"systemPrompt":                       g.SystemPrompt,
		"conversationTitle":                  g.ConversationTitle,
		"callError":                          g.CallError,
		"input.user.text":                    g.Input[0].Parts[0].Text,
		"input.assistant.text":               g.Input[1].Parts[0].Text,
		"input.assistant.thinking":           g.Input[1].Parts[1].Thinking,
		"input.assistant.toolCallInputJson":  string(g.Input[1].Parts[2].ToolCall.InputJSON),
		"input.tool.text":                    g.Input[2].Parts[0].Text,
		"input.tool.toolResultContent":       g.Input[2].Parts[1].ToolResult.Content,
		"input.tool.toolResultContentJson":   string(g.Input[2].Parts[1].ToolResult.ContentJSON),
		"output.assistant.text":              g.Output[0].Parts[0].Text,
		"output.assistant.thinking":          g.Output[0].Parts[1].Thinking,
		"output.assistant.toolCallInputJson": string(g.Output[0].Parts[2].ToolCall.InputJSON),
		"output.tool.text":                   g.Output[1].Parts[0].Text,
		"output.tool.toolResultContent":      g.Output[1].Parts[1].ToolResult.Content,
		"output.tool.toolResultContentJson":  string(g.Output[1].Parts[1].ToolResult.ContentJSON),
	}
}

func TestConformanceRedactionGenerationSlots(t *testing.T) {
	fixtures := loadGenerationFixtures(t)

	for _, tc := range fixtures.Cases {
		t.Run(tc.ID, func(t *testing.T) {
			sanitizer := newSecretRedactionSanitizer(
				func(string) (string, bool) { return "", false },
				SecretRedactionOptions{
					RedactInputMessages:  &tc.RedactInputMessages,
					RedactEmailAddresses: &tc.RedactEmailAddresses,
				},
			)

			sanitized := sanitizer(buildProbeGeneration(fixtures.Probe["input"]))
			got := generationSlotValues(sanitized)

			for slot, mode := range tc.Slots {
				want, ok := fixtures.Probe[mode]
				if !ok {
					t.Fatalf("slot %s: unknown mode %q", slot, mode)
				}
				actual, ok := got[slot]
				if !ok {
					t.Fatalf("slot %s is in the fixture but not built by the Go harness", slot)
				}
				if actual != want {
					t.Errorf("slot %s (mode %s)\n got: %q\nwant: %q", slot, mode, actual, want)
				}
			}

			for slot := range got {
				if _, ok := tc.Slots[slot]; !ok {
					t.Errorf("slot %s is built by the Go harness but missing from the fixture", slot)
				}
			}

			// Media is the one part kind the other three SDKs have no equivalent
			// for, so it stays out of the shared matrix. The sanitizer leaves it
			// alone; metadata-only capture strips media URLs instead.
			if url := sanitized.Input[0].Parts[1].Media.URL; url != fixtures.Probe["skip"] {
				t.Errorf("media url\n got: %q\nwant: %q", url, fixtures.Probe["skip"])
			}
		})
	}
}
